### PR TITLE
Deploy `dandidav` service to Heroku

### LINKED ISSUE #166: Expose dandiarchive as webdav service
To facilitate integration in various external projects (e.g., OSDF with [Pelican](https://pelicanplatform.org/) underneath) which can interface to webdav services.  Treat it as unification of our API to a standard API for file access.

TODOs:
- [x] decide on how to map our versions for dandisets within the path or some webdav supported mechanism for versioning (if there is any)
- [x] double-check if webdav supports "redirection" so we do not need to tunnel all bytes from S3 through the webdav service but rather just redirect.
- [x] implement mapping/endpoint: I wonder even if it somehow could be just the `/paths` endpoint, but likely not since that one is dandiset specific, but may be it could be at the top level of API
- [x] decide on implementation
   - could be internal to dandi-archive.  *pros:*   more efficient since would operate directly on DB/within Python, likely minimizing flood of the logs; it would come "included" by default with any dandi-archive installation; would not require another web server to provide it;  as our API still in 0.x and adds breakages from time to time -- would be easier to keep consistent; *cons:* it would be easier/faster to develop/fix as independent service likely
   - could be some additional external service. *pros:* faster/easier to develop; would stress test our API more to see what it is missing;  *cons:* solution would need to provide yet another webserver
- [ ] deploy as webdav.dandiarchive.org or at some other end-point [(discussion below)](https://github.com/dandi/dandi-infrastructure/issues/166)

may be it could be implemented as independent service, not part of API.  

attn @jwodder -- how much did you play with webdav?

edits:
- https://github.com/mar10/wsgidav looks very promising and already has (non-production) backends for various things, e.g. for mercurial: https://wsgidav.readthedocs.io/en/latest/addons-mercurial.html . 
- re structure - I think we at large should mimic our S3, just adding a little more of consistency and have `dandisets/` and `zarrs/` (not `zarr/`) folders (problem for `zarr/` though -- heavy folder listing).
- on the best way to expose versions, possibilities, assuming that we do have some global prefix which ends with `{dandiset_id}`.
  - following mercurial example above -- have `draft/`, `released/` (or `latest/` - most recent release), `releases/{VERSION}`.  It kinda makes it to require `draft/` or other prefix folder to get to content but it is consistent so I like it most.
  - have `draft/`, `released/` (or `latest/`) and then all versions at the same level. kinda ok, but since numbered releases would most likely sort first and oldest first -- I think it would be not that convenient of a default view... 
  - "smarty pants" version -- kinda making versioning optional and reacting to the regex for version at the top level and if not - giving the tree of the draft or most recent version... too smarty - don't like it
  - following our URL schema `dandi://INSTANCE/DANDISET_ID[@VERSION][/PATH]` and thus incorporating version into the folder name for dandiset  -- IMHO ugly 
- FWIW: filed https://github.com/mar10/wsgidav/issues/303

### PR DESCRIPTION
**Pull request by mvandenburgh**
Closes #166

### CODE REVIEW

**Comment by mvandenburgh** [comment]
@jwodder here's the Terraform code for deploying dandidav. The service itself is live at `webdav.dandiarchive.org`, and is currently pointed at [my fork](https://github.com/mvandenburgh/dandidav) of the repo. Once https://github.com/dandi/dandidav/pull/96 and https://github.com/dandi/dandidav/pull/95 are merged, we can point Heroku at the actual repo. 
Let me know if you have any questions about this PR, or want any more explanations/a walkthrough of the code here.

File: terraform/webdav.tf [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+resource "heroku_formation" "webdav_heroku_web" {
+  app_id   = heroku_app.webdav.id
+  type     = "web"
+  size     = "basic"
+  quantity = 1
+}
+
+# Enable the runtime-dyno-metadata feature for the Django app
+# https://devcenter.heroku.com/articles/dyno-metadata
```
Comment (danlamanna) [inline terraform/webdav.tf:23]: I'm guessing this is to make the sha available to the binary? IMO this is a what-not-why comment.

**Comment by jwodder** [comment]
@mvandenburgh 

* I've merged https://github.com/dandi/dandidav/pull/95.  Please update your fork so we can confirm it works.
* What exactly causes the project to be built using Rust?  Does the presence of the Rust buildpack in `webdav.tf` automatically cause `cargo build` to be run?
* Once this PR is merged, what will trigger deployment of the latest commit of `dandidav` to Heroku?  What would it take to automatically deploy whenever a PR is merged in `dandidav`?  Could PRs with certain labels be excluded from automatic deployment — or, alternatively, could only PRs with certain labels be deployed?

**Comment by mvandenburgh** [comment]
> * I've merged [Fetch Git commit hash from `HEROKU_SLUG_COMMIT` if Git info not available dandidav#95](https://github.com/dandi/dandidav/pull/95).  Please update your fork so we can confirm it works.

I updated my fork. I see that there's some text on the bottom of the page, `[dandidav](https://github.com/dandi/dandidav), v0.2.0, commit`. Is the commit hash supposed to be listed after that?

> * What exactly causes the project to be built using Rust?  Does the presence of the Rust buildpack in `webdav.tf` automatically cause `cargo build` to be run?

Yes, the Rust buildpack automatically runs `cargo build`. See https://github.com/emk/heroku-buildpack-rust/blob/master/bin/compile#L140

> * Once this PR is merged, what will trigger deployment of the latest commit of `dandidav` to Heroku?  What would it take to automatically deploy whenever a PR is merged in `dandidav`?  Could PRs with certain labels be excluded from automatic deployment — or, alternatively, could only PRs with certain labels be deployed?

Pushes to the `main` branch of the connected GitHub repo will trigger a deployment to Heroku. Heroku doesn't support conditional deployment based on labels. If that is a requirement, we could potentially adopt the same solution we use on dandi-archive (that is - use `auto` to cut a github release when a PR with a `release` label is merged, and then have a GitHub Action that runs on releases that manually compiles the Rust app and pushes the binary to Heroku, instead of using the Heroku GitHub integration).

File: terraform/webdav.tf (hunk 1 above)
Comment (mvandenburgh) [reply terraform/webdav.tf]: Yes. I adjusted that comment to explain the "why" instead.

**Comment by jwodder** [comment]
@mvandenburgh 

> Is the commit hash supposed to be listed after that?

Yes, it's supposed to look like https://dandi.centerforopenneuroscience.org.  The lack of a hash indicates that `HEROKU_SLUG_COMMIT` was not set at build time (or, theoretically, that it was set to the empty string) — though the bare "commit" still shouldn't be there when that happens (fixed in https://github.com/dandi/dandidav/pull/97).

**Comment by jwodder** [comment]
@mvandenburgh Are you looking into why `HEROKU_SLUG_COMMIT` didn't take effect?

**Comment by mvandenburgh** [comment]
> @mvandenburgh Are you looking into why `HEROKU_SLUG_COMMIT` wasn't set?

Yes. Well, I know why it wasn't set - the `HEROKU_SLUG_COMMIT` environment variable (as well as any other environment variables set in Heroku) is only set at runtime, i.e. when the compiled binary is executed. It is not set in the buildpack environment at compile time. 

I'm still investigating if there's a way to get it to work, but this may be another reason to consider the approach `dandi-archive` uses that I mentioned in [this comment](https://github.com/dandi/dandi-infrastructure/pull/167#issuecomment-1979876783).

**Pull request merged**